### PR TITLE
[FIX] base: do not delete records which are linked otherwise

### PR DESCRIPTION
**Pull request by wtaferner**
Only the last model data entry should delete a record if it exists.

**Description of the issue/feature this PR addresses:**
If model data is linking multiple times a record it might happen that the first deletion is destroying the rest of the links. So, how did it happen with Odoo Core? I was feeding the product and the template via data module entries (creating first a template and then linking all the products) and magically Odoo does add multiple links to the template and when I try to delete a single variant the next time I upgrade the module it will completely delete the template and therefore ALL products linked to it even if the template xml id still exists in the data files...

**Current behavior before PR:**
ALL variants get deleted if a single variant is triggering an update of a magic data xml id of the template

**Desired behavior after PR is merged:**
Only the variants get deleted and the data entries which are not existing anymore

@xmo-odoo 
As talked with @tde-banana-odoo this is a weird behavior and my suggestion is to make sure we only delete the record if the last xml id is about to be removed. Based on this we could also programatically protect in the database by adding an additional data entry.

What do you think? Let's discuss or I would be happy if you agree with the approach and the solution. 

Info: @wt-io-it

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by KangOl** [comment]
Not a such behavior change should be done in stable version.

**Comment by wtaferner** [comment]
@KangOl 
Kind of agree, although why should we accept that more is deleted than you would expect on a simple upgrade? So, the reason why I propose this is that it does fix a severe behavior on module upgrade.

I just pushed a better version of it with just doing a search_count instead of full search.

Did I miss any (business) use case where this is actually wanted behavior, then I would completely agree with changing it in master.

**Comment by KangOl** [comment]
Actually, IMD created for an _inherit records should be in noupdate.

**Comment by wtaferner** [comment]
So, all these entries created from Odoo magically for `product_product_*_val_attribute_x_product_template` in my module's namespace which are referencing the same entry I introduced to be updated via my modules xml id (`product_template_*`) should be `noupdate` and would therefore not delete the template on deletion of the `product.product` but only the model data entry? Right, I agree, this would work.

Hmm, this would have been smart and could be an active workaround on the database layer, but it does not save us from misbehaving old data wrongly created in the database if you are not aware of the issue on doing such a module upgrade and I indeed like the embedded feature that only the last removal will delete a record. So, we win two battles...one bug down and a useful feature added.

But this is more like a side story as I am really interested in use cases where this would be a change of wanted behavior that it is not applicable on stable? I am really for merging it in master if we have a safe fix for my issue in 11.0 (latest stable) too without knowing the details what was wrongly implemented before.

Do you know the code where this magic external id is created? I will provide another fix for that if you are short in time.

**Comment by KangOl** [comment]
https://github.com/odoo/odoo/blob/bb8ded772a857ad9d8f6d14270bc2917a27711d8/odoo/addons/base/ir/ir_model.py#L1443-L1450
https://github.com/odoo/odoo/blob/bb8ded772a857ad9d8f6d14270bc2917a27711d8/odoo/addons/base/ir/ir_model.py#L1486-L1492

**Comment by wtaferner** [comment]
@KangOl 
After testing your approach with `noupdate` in the wild I can confirm that it resolves the issue, but the downside is that useless (as the real meaning/context is lost when the product is removed via upgrade) external ids remain in the system until the module is uninstalled or the external ids are removed at some point manually. Also it locks away these names for using it for other objects by coincidence as it will just ignore module data by "non-module" data ;-)

So, my conclusion is that the approach of mine is the cleanest way to handle the issue consistently and keep the model data as clean as possible. I hope you appreciate the research I did based on your input and see the advantage of my solution.

So, basically if there are no business use cases which would rely on the current behavior I would appreciate that this is merged in stable (11.0). 

@xmo-odoo
Do you have technically any objections or suggestions on the way I tried to get rid of the issue? 

@nim-odoo 
Any objections from your point of view in terms of business use cases or wanted behavior which would imply a change in behavior which would speak against merging this in latest stable?

**Comment by xmo-odoo** [comment]
Heya @wtaferner after talking about this issue with @odony and @KangOl in person, while the bit about xids is odd its actual impact in the general case should be limited as it's not possible to get multiple module-linked xids for the same record safe by doing so very explicitly or:

> magically Odoo does add multiple links to the template

namely that when you create a record which _inherits from an other and provide the parent, the creation of the child will *still* add a new XID to the parent, leading to its unintended deletion when the child is removed. @KangOl's comments were somewhat misguided but still what we think the right area of concern: the issue should probably be fixed in _update and dealt with by not adding these extraneous xids for the existing explicitly created parent.

**Comment by wtaferner** [comment]
@xmo-odoo 
Sorry for coming back late to your answer. I think you got the issue, but not providing any solution for stable where it is very dangerous as said, is kind of "lazy".
My solution might be not perfect, but does avoid the wrong deletion. Of course we could add an explicit...DO NOT FORWARD-PORT (NOT A FEATURE) comment that it is clear for everyone and provide the real solution on 11.5 (12.0), but what I do not get about such discussed and confirmed PR/issues that nothing is done to resolve it where it happens (in stable)

And, yet it is close to the "new" stable and still there is no solution around the corner, right? I hope you do understand my PoV.

**Comment by wtaferner** [comment]
Closing as this was fixed (in 12.0) with https://github.com/odoo/odoo/commit/91882dd97238a006ef711b33a96ba88acb862847

**Comment by wtaferner** [comment]
Ok, to fast, fix in another context, so it is still not fixed. OPW was closed that it will be fixed in the future version.

**Comment by fgi-odoo** [comment]
@wtaferner I do a functional check on this in order to move forward. I didn't manage to reproduce it with templates and variants.
- I created a product template,
- created some variants (using the attribute table)
- deleted a variant
- update the module sale
- the template and all the other variants are still there...

what is missing here? thank you.

**Comment by wtaferner** [comment]
@fgi-odoo 
I think there is nothing you could do in terms of functional as you need a module maintained product to be created on module install.

**Comment by Elkasitu** [comment]
@wtaferner How is it possible that https://github.com/odoo/odoo/commit/91882dd97238a006ef711b33a96ba88acb862847 does not fix your use-case but your patch does? What do you mean by different context? I fail to see how the patches are different in principle, they should both produce the same end result.

Also it'd be great if you could detail the steps to reproduce your bug, thanks

**Comment by pedrobaeza** [comment]
Coming to this from codetriage.com. Is this still relevant?

**Comment by wtaferner** [comment]
@pedrobaeza 
It is, unfortunately I had no time to test the incomplete fix yet and report back to @Elkasitu 
I think we said that I should try to remove the ```("id", "not in", bad_imd_ids)``` from the domain and test against my use case which would then replace my proposed fix.

**Comment by pedrobaeza** [comment]
Then should we close it?

**Comment by wtaferner** [comment]
Well, I will easier forget about it, but I close it now and will reopen in case I have time to present new solutions and evidence (easy reproduction in less than 2 minutes for everyone)
